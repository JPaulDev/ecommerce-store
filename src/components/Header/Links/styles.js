import styled from 'styled-components';

export const Container = styled.div`
  column-gap: 40px;
  display: flex;
  justify-content: flex-end;
`;

export const Link = styled.a`
  align-items: center;
  color: var(--color-grey-3);
  column-gap: 10px;
  display: flex;
  text-decoration: none;
  transition: color 200ms ease-in-out;

  &:hover {
    color: white;
  }
`;

export const TopText = styled.span`
  color: white !important;
  display: block;
`;

export const Text = styled.span`
  font-size: 0.8rem;
`;
