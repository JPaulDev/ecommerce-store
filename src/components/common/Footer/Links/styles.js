import styled from 'styled-components';

export const List = styled.ul`
  list-style: none;
  padding: 0 20px;

  li {
    padding: 7px 0;
  }

  @media screen and (min-width: 650px) {
    padding: 0 10px;
  }

  @media screen and (min-width: 860px) {
    padding: 0;
  }
`;

export const Link = styled.a`
  color: var(--color-grey-4);
  font-size: var(--font-size-13);
  text-decoration: none;

  &:hover {
    text-decoration: underline;
  }
`;
