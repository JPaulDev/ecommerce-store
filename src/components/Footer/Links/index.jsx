import { useEffect, useState } from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import useMediaQuery from '../../../hooks/useMediaQuery';
import MaterialIcon from '../../MaterialIcon';
import { Heading, DropdownButton } from '../styles';
import * as Styled from './styles';

const MotionList = motion(Styled.List);

const parentVariants = {
  closed: {
    height: 0,
    transition: {
      ease: 'easeInOut',
      duration: 0.5,
      when: 'afterChildren',
    },
  },
  open: {
    height: 'auto',
    transition: {
      ease: 'easeInOut',
      duration: 0.5,
      when: 'beforeChildren',
    },
  },
};

const childVariants = {
  closed: (index) => ({
    opacity: 0,
    x: -50,
    transition: {
      ease: 'easeInOut',
      delay: 0.5 - index * 0.1,
      duration: 0.2,
    },
  }),
  open: (index) => ({
    opacity: 1,
    x: 0,
    transition: {
      ease: 'easeInOut',
      delay: index * 0.1,
      duration: 0.2,
    },
  }),
};

export default function Links({ item }) {
  const [isOpen, setIsOpen] = useState(window.innerWidth >= 860);
  const isMatch = useMediaQuery('(min-width: 860px)');

  const handleOpen = () => setIsOpen(!isOpen);

  // Prevents the menu being stuck closed when resizing from a smaller window width
  useEffect(() => {
    isMatch ? setIsOpen(true) : setIsOpen(false);
  }, [isMatch]);

  return (
    <div>
      {isMatch ? (
        <Heading>{item.heading}</Heading>
      ) : (
        <DropdownButton
          type="button"
          isOpen={isOpen}
          aria-expanded={isOpen}
          aria-controls={item.heading}
          onClick={handleOpen}
        >
          {item.heading}
          <MaterialIcon iconSize="30px">chevron_right</MaterialIcon>
        </DropdownButton>
      )}
      <AnimatePresence initial={false}>
        {isOpen && (
          <MotionList
            id={item.heading}
            variants={parentVariants}
            initial="closed"
            animate="open"
            exit="closed"
          >
            {item.links.map((link, index) => (
              <motion.li
                key={link.text}
                custom={index}
                variants={childVariants}
              >
                <Styled.Link href={link.url}>{link.text}</Styled.Link>
              </motion.li>
            ))}
          </MotionList>
        )}
      </AnimatePresence>
    </div>
  );
}
