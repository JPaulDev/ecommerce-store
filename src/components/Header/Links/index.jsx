import useMediaQuery from '../../../hooks/useMediaQuery';
import { ReactComponent as AccountIcon } from '../../../assets/svg/account.svg';
import { ReactComponent as BasketIcon } from '../../../assets/svg/basket.svg';
import * as Styled from './styles';

export default function Links() {
  const isMatch = useMediaQuery('(min-width: 575px)');

  return (
    <Styled.Container>
      <Styled.Link href="" aria-label="Account" title="Account">
        <AccountIcon width="32px" height="30px" fill="var(--color-orange-1)" />
        {isMatch && (
          <div>
            <Styled.TopText>Account</Styled.TopText>
            <Styled.Text>Sign in | Register</Styled.Text>
          </div>
        )}
      </Styled.Link>
      <Styled.Link href="" aria-label="Basket" title="Basket">
        <BasketIcon width="34px" height="30px" fill="var(--color-orange-1)" />
        {isMatch && (
          <div>
            <Styled.TopText>Basket</Styled.TopText>
            <Styled.Text>Subtotal Â£0.00</Styled.Text>
          </div>
        )}
      </Styled.Link>
    </Styled.Container>
  );
}
