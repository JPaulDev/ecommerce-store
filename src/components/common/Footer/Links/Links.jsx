import Link from 'next/link';
import { useEffect, useState } from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import useMediaQuery from '../../../../lib/hooks/useMediaQuery';
import { ChevronRight } from '../../../icons';
import { Heading, DropdownButton } from '../styles';
import * as Styled from './styles';

const parentVariants = {
  open: {
    height: 'auto',
    transition: {
      ease: 'easeInOut',
      duration: 0.5,
      when: 'beforeChildren',
    },
  },
  closed: {
    height: 0,
    transition: {
      ease: 'easeInOut',
      duration: 0.5,
      when: 'afterChildren',
    },
  },
};

const childVariants = {
  open: (index) => ({
    opacity: 1,
    x: 0,
    transition: {
      ease: 'easeInOut',
      delay: index * 0.1,
      duration: 0.2,
    },
  }),
  closed: (index) => ({
    opacity: 0,
    x: -50,
    transition: {
      ease: 'easeInOut',
      delay: 0.5 - index * 0.1,
      duration: 0.2,
    },
  }),
};

export default function Links({ item }) {
  const [isOpen, setIsOpen] = useState(false);
  const isMatch = useMediaQuery('(min-width: 860px)');

  const handleOpen = () => setIsOpen(!isOpen);

  // Prevents the menu being stuck closed when resizing from a smaller window
  // width.
  useEffect(() => {
    isMatch ? setIsOpen(true) : setIsOpen(false);
  }, [isMatch]);

  return (
    <div>
      {isMatch ? (
        <Heading>{item.heading}</Heading>
      ) : (
        <DropdownButton
          as="button"
          type="button"
          isOpen={isOpen}
          aria-expanded={isOpen}
          aria-controls={item.heading}
          onClick={handleOpen}
        >
          {item.heading}
          <ChevronRight width={15} height={20} />
        </DropdownButton>
      )}
      <AnimatePresence>
        {isOpen && (
          <Styled.List
            as={motion.ul}
            id={item.heading}
            variants={parentVariants}
            initial="closed"
            animate="open"
            exit="closed"
          >
            {item.links.map((link, index) => (
              <motion.li
                key={link.text}
                custom={index}
                variants={childVariants}
              >
                <Link href={link.href} passHref>
                  <Styled.Link>{link.text}</Styled.Link>
                </Link>
              </motion.li>
            ))}
          </Styled.List>
        )}
      </AnimatePresence>
    </div>
  );
}
