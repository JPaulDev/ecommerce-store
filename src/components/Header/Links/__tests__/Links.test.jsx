import { screen, render } from '@testing-library/react';
import Links from '../index';

const linkText = [
  /account$/i,
  /sign in | register/i,
  /basket$/i,
  /subtotal Â£0.00/i,
];

describe('Links component at different screen widths', () => {
  it('renders links with their icons and text visible at 575px and above', () => {
    window.resizeTo(575, 768);
    render(<Links />);

    linkText.forEach((item) => {
      expect(screen.getByText(item)).toBeInTheDocument();
    });
  });

  it('renders links with only their icons visible at 574px and below', () => {
    window.resizeTo(574, 768);
    render(<Links />);

    linkText.forEach((item) => {
      expect(screen.queryByText(item)).not.toBeInTheDocument();
    });
  });
});
