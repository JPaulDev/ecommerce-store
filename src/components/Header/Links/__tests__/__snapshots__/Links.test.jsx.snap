// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Links component at a screen width of 574px and below renders links with no text and only their icons visible 1`] = `
<div
  className="sc-bczRLJ ereDqo"
>
  <a
    aria-label="Account"
    className="sc-gsnTZi cVgHpq"
    href=""
    title="Account"
  >
    <svg
      fill="var(--color-orange-1)"
      height={30}
      width={32}
    >
      account.svg
    </svg>
  </a>
  <a
    aria-label="Basket"
    className="sc-gsnTZi cVgHpq"
    href=""
    title="Basket"
  >
    <svg
      fill="var(--color-orange-1)"
      height={30}
      width={34}
    >
      basket.svg
    </svg>
  </a>
</div>
`;

exports[`Links component at a screen width of 575px and above renders links with their icons and text visible 1`] = `
<div
  className="sc-bczRLJ ereDqo"
>
  <a
    aria-label="Account"
    className="sc-gsnTZi cVgHpq"
    href=""
    title="Account"
  >
    <svg
      fill="var(--color-orange-1)"
      height={30}
      width={32}
    >
      account.svg
    </svg>
    <div>
      <span
        className="sc-dkzDqf ghjqcX"
      >
        Account
      </span>
      <span
        className="sc-hKMtZM hMHOua"
      >
        Sign in | Register
      </span>
    </div>
  </a>
  <a
    aria-label="Basket"
    className="sc-gsnTZi cVgHpq"
    href=""
    title="Basket"
  >
    <svg
      fill="var(--color-orange-1)"
      height={30}
      width={34}
    >
      basket.svg
    </svg>
    <div>
      <span
        className="sc-dkzDqf ghjqcX"
      >
        Basket
      </span>
      <span
        className="sc-hKMtZM hMHOua"
      >
        Subtotal Â£0.00
      </span>
    </div>
  </a>
</div>
`;
